che-core-vfs-impl/src/main/java/org/eclipse/che/vfs/impl/fs/FSMountPoint.java

```diff
@@ -1445,43 +1417,54 @@ void updateACL(VirtualFileImpl virtualFile, List<AccessControlEntry> acl, boolea
     }
 
 
-    private boolean hasPermission(VirtualFileImpl virtualFile, String p, boolean checkParent) {
-        final VirtualFileSystemUser user = userContext.getVirtualFileSystemUser();
+    protected boolean hasPermission(VirtualFileImpl virtualFile, BasicPermissions p, boolean checkParent) {
+        // Find the closes access control list on the line of ancestry, starting with the virtual file itself
         Path path = virtualFile.getVirtualFilePath();
-        while (path != null) {
-            final AccessControlList accessControlList = aclCache[path.hashCode() & MASK].get(path);
-            if (!accessControlList.isEmpty()) {
-                final Principal userPrincipal = DtoFactory.getInstance().createDto(Principal.class)
-                                                          .withName(user.getUserId()).withType(Principal.Type.USER);
-                Set<String> userPermissions = accessControlList.getPermissions(userPrincipal);
-                if (userPermissions != null) {
-                    return userPermissions.contains(p) || userPermissions.contains(BasicPermissions.ALL.value());
-                }
-                Collection<String> groups = user.getGroups();
-                if (!groups.isEmpty()) {
-                    for (String group : groups) {
-                        final Principal groupPrincipal = DtoFactory.getInstance().createDto(Principal.class)
-                                                                   .withName(group)
-                                                                   .withType(Principal.Type.GROUP);
-                        userPermissions = accessControlList.getPermissions(groupPrincipal);
-                        if (userPermissions != null) {
-                            return userPermissions.contains(p) || userPermissions.contains(BasicPermissions.ALL.value());
-                        }
-                    }
-                }
-                final Principal anyPrincipal = DtoFactory.getInstance().createDto(Principal.class)
-                                                         .withName(VirtualFileSystemInfo.ANY_PRINCIPAL)
-                                                         .withType(Principal.Type.USER);
-                userPermissions = accessControlList.getPermissions(anyPrincipal);
-                return userPermissions != null && (userPermissions.contains(p) || userPermissions.contains(BasicPermissions.ALL.value()));
+        AccessControlList accessControlList;
+        while (true) {
```

Comment:
the loop body has multiple terminal points which is somewhat restrictive because i intend to reduce double checks to zero. That's why I wrote the loop as simple as possible with verbose documentation ...